### DISCUSSION TITLE
Rate Limit on add_dockerfile()

### DISCUSSION THREAD

**Discussion by aftonsteps** [ThinkR-open/golem#570] [Q&A] [answered]
Trying to run `golem::add_dockerfile()` results in:

```
Error: HTTP error 401.
    Bad credentials.

    Rate limit remaining: 59/60
    Rate limit reset at: 2020-12-11 05:05:09 UTC
```

Searching around, I couldn't find anyone with the same issue. I hesitate to open an issue because this seems like a 'me' problem and not a 'golem' problem. I didn't realize the `add_dockerfile()` function required credentials, though, and I'm not sure where to find an entry point to provide them.

Any help is appreciated!

**Comment by ColinFay** [comment, marked as answer]
Hey, 

So this is due to the fact that there is a rate limit when trying to install packages from GitHub without being logged (you can only make 60 requests by hour). 

It's not purely `{golem}` related, but can indeed arise if you're rebuilding your Dockerfile too many time (as you will reach the GitHub API limit). That's the reason why `install_github()` has a `auth_token` parameter: https://remotes.r-lib.org/reference/install_github.html.

Two solutions here: 

- Wait another hour for the rate limit to restart 
- Add an oauth token to `install_github` inside your `Dockerfile`. You can create one automatically by calling `usethis::browse_github_pat()`. 

Hope this helps, 

Colin

**Reply by aftonsteps** [reply]
Ah, thank you! I didn't realize that `add_dockerfile()` installed anything from GitHub. I'm still not sure what the issue really is, though, because I am using `GITHUB_PAT` and have been for a while. On the other hand, when I tried not using `GITHUB_PAT` via `Sys.unsetenv("GITHUB_PAT")`, that resolved it, and I could make a Dockerfile, which doesn't make a lot of sense to me if `add_dockerfile()` does require an auth token. 😅

**Reply by ColinFay** [reply]
Ok so I might have misunderstood the issue, is your problem appearing just after `add_dockerfile()`, or is it when you build the container?

**Reply by aftonsteps** [reply]
It's just when running `add_dockerfile()`. When I run that function, the message `i Please wait while we compute system requirements...` appears, and then I get the rate limit error. 🙁  However, the issue is resolved by unsetting `GITHUB_PAT` (although I don't understand why).

**Reply by fmmattioni** [reply]
I guess I am familiar with the issue.. I guess the problem is that, even though you are not installing the dev version from github, `remotes::install_version()` or even `remotes::install_local()` trigger this error. I always used `RUN echo "GITHUB_PAT=..." >> $HOME/.Renviron` at the top of the `Dockerfile` since then. I have no idea what is the root of the problem, though.

**Reply by aftonsteps** [reply]
Oh hmmmm, I see. And this issue arises when you even try to create the Dockerfile (not build the Docker image), via something like `golem::add_dockerfile()`? I wasn't even able to create a Dockerfile, so adding a line to the top of the Dockerfile would not be possible for me. 😅

**Reply by fmmattioni** [reply]
Oh ok... yeah, I never had it while creating the file.. only when building the image
